ampd/src/handlers/solana_verify_msg.rs

```diff
@@ -0,0 +1,433 @@
+use async_trait::async_trait;
+use cosmrs::cosmwasm::MsgExecuteContract;
+use error_stack::ResultExt;
+use serde::Deserialize;
+use solana_transaction_status::EncodedConfirmedTransactionWithStatusMeta;
+use std::collections::HashSet;
+use std::convert::TryInto;
+use tracing::info;
+
+use axelar_wasm_std::voting::{PollId, Vote};
+use events::{Error::EventTypeMismatch, Event};
+use events_derive::try_from;
+use tokio::sync::watch::Receiver;
+use voting_verifier::msg::ExecuteMsg;
+
+use crate::event_processor::EventHandler;
+use crate::handlers::errors::Error;
+use crate::queue::queued_broadcaster::BroadcasterClient;
+use crate::solana::{json_rpc::SolanaClient, verifier::verify_message};
+use crate::types::TMAddress;
+
+type Result<T> = error_stack::Result<T, Error>;
+
+#[derive(Deserialize, Debug, PartialEq)]
+pub struct Message {
+    pub tx_id: String,
+    pub event_index: u64,
+    pub destination_address: String,
```

Comment:
We dont actually know of what blockchain the `destination_address` will be. Do you mean the `source_address` field ?

Reply:
Aha, so i see this type `PubKeyWrapper` , [it implements](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/pub_key_wrapper.rs#L27-L38) the `borsh` serialization traits

```rust

pub struct PubkeyWrapper(Pubkey);

impl From<Pubkey> for PubkeyWrapper {
    fn from(value: Pubkey) -> Self {
        Self(value)
    }
}

impl Deref for PubkeyWrapper {
    type Target = Pubkey;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl BorshDeserialize for PubkeyWrapper {
    fn deserialize_reader<R: std::io::Read>(reader: &mut R) -> std::io::Result<Self> {
        let inner: [u8; 32] = <[u8; 32]>::deserialize_reader(reader)?;
        Ok(PubkeyWrapper(inner.into()))
    }
}

impl BorshSerialize for PubkeyWrapper {
    fn serialize<W: std::io::Write>(&self, writer: &mut W) -> std::io::Result<()> {
        <[u8; 32]>::serialize(&self.0.to_bytes(), writer)
    }
}
```

But, i am unsure whats the utility of using that trait in the message verification part. Maybe for comparisons ? let me check.

Reply:
I have found a couple of things:

1. I think we should be using the original solana GatewayEvent::CallContract variant from our `solana-axelar` repo. [Not this one](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L15-L41).
2. We are actually [ignoring](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L47-L48) the [sender](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L22) field in event comparison. Which i think shouild match with the `source_address` .

Probably by fixing the above points, the use of `PubKeyWrapper` would make more sense ? I am going to fix the above points in order.

Reply:
`1` was fixed at 5236087946978511c6cec4b1e3fc2a95f6b349a9 . Hope the `unimplemented!()` code its not too hacky , let me know.

Reply:
> We dont actually know from what blockchain the destination_address will be. Do you mean the source_address field ?

Source, yes. Sorry about that. If string works and it's too much work to implement the other one, leave it. PubKeyWrapper was just a suggestion. I don't remember where it came from, but seems like it can be used to interpret a solana address 🤔 

> I think we should be using the original solana GatewayEvent::CallContract variant from our solana-axelar repo. [Not this one](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L15-L41).

Make sense.

> We are actually [ignoring](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L47-L48) the [sender](https://github.com/ctoyan/axelar-amplifier-fork/blob/484e3a4c30157696679548a518f58a3d5cd83c27/ampd/src/solana/verifier.rs#L22) field in event verification. Which i think shouild match with the source_address .

I think this is also correct. I believe I made a mistake in the relayer when constructing the axelar msg and assigned the solana GW address as `source_address`, where it should've been the sender. I'll open a on line PR for it.

> Hope the unimplemented!() code its not too hacky

If it's a common Rust pattern, feel free to leave it, but wouldn't it panic, which is not desirable, unless we really have to panic?

Reply:
> If it's a common Rust pattern, feel free to leave it, but wouldn't it panic, which is not desirable, unless we really have to panic?

Sorry you are right. Taking into account we are comparing a very specific type with a dedicated enum variant, i think we can just return "false" for the rest of Variants. Thanks ! fixed at 73841642c66cfcfa0b386d3095730cc35abe2355

Reply:
> I think this is also correct. I believe I made a mistake in the relayer when constructing the axelar msg and assigned the solana GW address as source_address, where it should've been the sender. I'll open a on line PR for it.

Now we added it to verification at 70d0fb416ee73fb2d563166b40efa4ecc402b6e0 :+1:

Reply:
> If string works and it's too much work to implement the other one, leave it. PubKeyWrapper was just a suggestion. 

The problem with using `PubKeyWrapper` as part of the [message](https://github.com/ctoyan/axelar-amplifier-fork/blob/0b246c67e0f4e4e82e1a127bbd637d476954155f/ampd/src/handlers/solana_verify_msg.rs#L24-L34) is that is not a base type that serde supports. As the parent struct implements `Deserialize` from serde, all the childs should also implement it or being part of the serde [data model](https://serde.rs/data-model.html). But in the case of  `PubKeyWrapper` its not the case. I think we would [need to tell serde how to deserialize that custom type](https://serde.rs/impl-deserialize.html)  . But, i do not think it worths it for now. We do string comparison.

```rust

#[derive(Deserialize, Debug, PartialEq)]  // this implements Deserialize
pub struct Message {
    pub tx_id: String,
    pub event_index: u64,
    pub destination_address: String,
    pub destination_chain: connection_router::state::ChainName,
    pub source_address: PubKeyWrapper, /// this would need to implement Deserialize
    #[serde(with = "axelar_wasm_std::hex")]
    pub payload_hash: [u8; 32],
}
```

Reply:
BTW remove the PubKeyWrapper copied code at 4e69e91f3ee0434857010a3fb0ff921a14c2b617 , as we are using the original now :+1: